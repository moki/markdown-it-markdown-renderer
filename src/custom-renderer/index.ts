import Renderer from 'markdown-it/lib/renderer';
import Token from 'markdown-it/lib/token';
import {Options} from 'markdown-it';

export type CustomRendererParams<S = {}> = {
    handlers?: CustomRendererHanlders;
    rules?: Renderer.RenderRuleRecord;
    mode?: CustomRendererMode;
    initState?: () => S;
};

export type CustomRendererHanlders = Record<string, Renderer.RenderRule | Renderer.RenderRule[]>;

export enum CustomRendererMode {
    Production,
    Development,
}

class CustomRenderer<State = {}> extends Renderer {
    protected mode: CustomRendererMode;
    protected handlers: Map<string, Renderer.RenderRule[]>;
    protected state: State;

    constructor({
        mode = CustomRendererMode.Production,
        handlers = {},
        rules = {},
        initState = () => ({} as State),
    }: CustomRendererParams<State>) {
        super();

        this.mode = mode;
        this.setRules(rules);

        this.state = initState();
        this.handlers = new Map<string, Renderer.RenderRule[]>();
        this.setHandlers({...handlers});
    }

    setRules(rules: Renderer.RenderRuleRecord) {
        for (const [name, rule] of Object.entries(rules)) {
            if (!rule || !name?.length) {
                continue;
            }

            this.rules[name] = rule.bind(this);
        }
    }

    setHandlers(rules: CustomRendererHanlders) {
        for (const [name, handler] of Object.entries(rules)) {
            if (!handler || !name?.length) {
                continue;
            }

            this.handle(name, handler);
        }
    }

    handle(type: string, handler: Renderer.RenderRule | Renderer.RenderRule[]) {
        const handlers = this.handlers.get(type) ?? [];

        const normalized = (Array.isArray(handler) ? handler : [handler]).map((h) => h.bind(this));

        this.handlers.set(type, [...handlers, ...normalized]);
    }

    render(tokens: Token[], options: Options, env: unknown) {
        let rendered = '';

        let children;
        let type;
        let len;
        let i;

        for (i = 0, len = tokens.length; i < len; i++) {
            type = tokens[i].type;
            children = tokens[i].children;

            if (type === 'inline' && Array.isArray(children)) {
                rendered += this.renderInline(children, options, env);

                continue;
            }

            rendered += this.processToken(tokens, i, options, env);
        }

        return rendered;
    }

    renderInline(tokens: Token[], options: Options, env: unknown) {
        let rendered = '';

        let len;
        let i;

        for (i = 0, len = tokens.length; i < len; i++) {
            rendered += this.processToken(tokens, i, options, env);
        }

        return rendered;
    }

    processToken(tokens: Token[], i: number, options: Options, env: unknown) {
        let rendered = '';

        const type = tokens[i].type;
        const handlers = this.handlers.get(type);
        const rule = this.rules[type];

        if (handlers) {
            for (const handler of handlers) {
                rendered += handler(tokens, i, options, env, this);
            }
        }

        if (rule) {
            rendered += rule(tokens, i, options, env, this);
        } else {
            rendered += this.renderToken(tokens, i, options);
        }

        return rendered;
    }
}

export {CustomRenderer};
export default {CustomRenderer};
