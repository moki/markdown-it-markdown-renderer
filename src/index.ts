export * from 'src/renderer';
export * from 'src/plugin';
