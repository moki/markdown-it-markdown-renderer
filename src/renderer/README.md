# renderer

`MarkdownRenderer` extends `markdown-it` [Renderer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/markdown-it/lib/renderer.d.ts) overwriting default behaviour with [custom rules]../rules/README.md)

## Usage

`MarkdownRenderer` can be parametrized with custom rules or in which mode it should be instantiated

### Constructor parameters

```
export type MarkdownRendererParams = {
    customRules?: Renderer.RenderRuleRecord;
    mode?: MarkdownRendererMode;
};
```

#### mode

`MarkdownRenderer` has two instance modes `MarkdownRendererMode.Production`(default one) and `MarkdownRendererMode.Development`

`MarkdownRendererMode.Development` - provides debug information about rules being called, in the form: `rule <rule_name> called with args <arguments_list>`

#### customRules

`markdown-it` [RenderRuleRecord](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/92075da4b8f7cf6b2bd5ba0299c4337341a56410/types/markdown-it/lib/renderer.d.ts#L7) object
