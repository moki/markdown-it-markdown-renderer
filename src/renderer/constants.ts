const NEW_LINE = '\n';
const ONE_SPACE = ' ';

export {NEW_LINE, ONE_SPACE};
export default {NEW_LINE, ONE_SPACE};
