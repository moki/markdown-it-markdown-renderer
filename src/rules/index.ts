import {code} from './inline/code';
import {emphasis} from './inline/emphasis';
import {hardbreak} from './inline/hardbreak';
import {image} from './inline/image';
import {link} from './inline/link';
import {softbreak} from './inline/softbreak';
import {text} from './inline/text';

import {paragraph} from './block/paragraph';

export default {
    ...code,
    ...emphasis,
    ...hardbreak,
    ...image,
    ...link,
    ...softbreak,
    ...text,
    ...paragraph,
};
