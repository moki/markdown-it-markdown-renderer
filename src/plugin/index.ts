import type MarkdownIt from 'markdown-it';
import {MarkdownRenderer, MarkdownRendererParams} from 'src/renderer';

// helpers
// always evaluate to provided value <v>
const always = (v: boolean) => () => v;
// always return input value <a> as a result
const id = (a: string) => a;

function mdRenderer(parser: MarkdownIt, parameters?: MarkdownRendererParams) {
    // disable escape rule
    parser.inline.ruler.at('escape', always(false));
    // disable entity rule
    parser.inline.ruler.at('entity', always(false));
    // disable links normalization
    parser.normalizeLink = id;

    const renderer = new MarkdownRenderer(parameters);

    // @ts-ignore
    parser.renderer = renderer;
}

export {mdRenderer};
export default {mdRenderer};
