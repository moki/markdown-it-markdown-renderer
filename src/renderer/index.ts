import Renderer from 'markdown-it/lib/renderer';
import Token from 'markdown-it/lib/token';

import {CustomRendererMode, CustomRenderer, CustomRendererParams} from 'src/custom-renderer';
import {inline} from 'src/rules/inline';
import {block} from 'src/rules/block';
import {renderEmptyBlockquote, renderBlockquote} from 'src/rules/block/blockquote';
import {renderEmptyListItem, renderOrderedList, renderUnorderedList} from 'src/rules/block/list';
import {
    sameRow,
    aligned,
    ContainerRenderer,
    Container,
    ContainerBase,
} from 'src/rules/block/containers';

import {NEW_LINE, ONE_SPACE} from './constants';

export const MarkdownRendererMode = CustomRendererMode;

export type MarkdownRendererParams<
    T = {},
    CT extends ContainerBase = ContainerBase,
> = CustomRendererParams<T> & {
    constants?: {
        EOL?: string;
        SPACE?: string;
    };
    containerRenderers?: ContainerRenderer<CT>[];
};

// Renderer Environment
// passed to every rule handling tokens render
export type MarkdownRendererEnv = {
    // array of original markdown string split by new lines
    // some rules are using it to achive more accurate rendering
    source?: string[];
};

class MarkdownRenderer<T = {}, CT extends ContainerBase = ContainerBase> extends CustomRenderer<T> {
    static defaultRules: Renderer.RenderRuleRecord = {...inline, ...block};
    static defaultContainerRenderers: Array<ContainerRenderer<ContainerBase>> = [
        renderEmptyBlockquote,
        renderEmptyListItem,
        renderOrderedList,
        renderUnorderedList,
        renderBlockquote,
    ];

    // constants
    // end of line used by renderer
    protected EOL: string;
    // space used by renderer
    protected SPACE: string;

    // pending tokens stack
    protected pending: Token[];

    // containers
    protected containers: Array<Container<CT>>;
    protected containerRenderers: Array<ContainerRenderer<CT>>;

    constructor(params: MarkdownRendererParams<T, CT> = {}) {
        const {
            initState,
            handlers,
            rules,
            containerRenderers = [],
            mode,
            constants: {EOL = NEW_LINE, SPACE = ONE_SPACE} = {},
        } = params;

        super({rules: {...MarkdownRenderer.defaultRules, ...rules}, handlers, initState, mode});

        this.EOL = EOL;
        this.SPACE = SPACE;

        this.pending = [];
        this.containers = new Array<Container<CT>>();

        this.containerRenderers = [
            ...MarkdownRenderer.defaultContainerRenderers,
            ...containerRenderers,
        ].map((rfn: ContainerRenderer<CT>) => rfn.bind(this));
    }

    // eslint-disable-next-line complexity
    renderContainer(caller: Token): string {
        let rendered = '';

        if (!caller) {
            throw new Error('provide caller token');
        }

        if (!this.containers.length) {
            return rendered;
        }

        for (let i = 0; i < this.containers.length; i++) {
            const container = this.containers[i];
            const previous = this.containers[i - 1];

            // common behaviour between containers
            // merge left trailling spaces and right leading spaces
            if (i && sameRow(previous, container) && aligned(previous, container)) {
                const lspaces = Math.max(container.lspaces - previous.tspaces, 0);

                this.containers[i].lspaces = lspaces;
            }

            const lspaces = this.containers[i].lspaces;

            // common behaviour render leading spaces
            rendered += this.SPACE.repeat(lspaces);

            for (const renderer of this.containerRenderers) {
                const rv = renderer(this.containers, i, caller);

                if (rv?.length) {
                    rendered += rv;
                    break;
                }
            }

            // common behaviour between containers
            // mark container as rendered
            this.containers[i].rendered = true;
        }

        // returns result of running one of the handlers
        return rendered;
    }
}

export {MarkdownRenderer};
export default {MarkdownRenderer};
