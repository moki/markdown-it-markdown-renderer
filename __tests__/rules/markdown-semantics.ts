import MarkdownIt from 'markdown-it';
import {MarkdownRenderer, MarkdownRendererEnv, MarkdownRendererMode} from 'src/renderer';
import {tests} from 'commonmark-spec';

import {CommonMarkSpecEntry} from './__fixtures__';

const renderer = new MarkdownRenderer({mode: MarkdownRendererMode.Production});

const md = new MarkdownIt('commonmark', {html: true});

// helpers
// always evaluate to provided value <v>
const always = (v: boolean) => () => v;
// always return input value <a> as a result
const id = (a: string) => a;

// modify markdown-it parser behaviour
// disable escape rule
md.inline.ruler.at('escape', always(false));
// disable entity rule
md.inline.ruler.at('entity', always(false));
// disable links normalization
md.normalizeLink = id;

// @ts-ignore
md.renderer = renderer;

const semanticsTests = new Set([
    253, 254, 255, 256, 258, 259, 260, 262, 263, 272, 275, 276, 277, 279, 280, 306, 307, 308, 309,
    312, 314, 315, 316, 317, 319, 325, 326,
]);

const units = tests.filter(({number}) => semanticsTests.has(number));

// omit first and last empty lines
const filterOutEmptyFstLst = (line, i, ls) => (i && i + 1 !== ls.length) || line.trim().length;
const normalizeMD = (str: string) => str.split('\n').filter(filterOutEmptyFstLst).join('\n');

describe('markdown semantics', () => {
    units.forEach((entry: CommonMarkSpecEntry) => {
        const {section, number, markdown} = entry;

        const name = `${section} ${number}`;

        test(name, () => {
            // passing array of original markdown string split by new lines
            const env: MarkdownRendererEnv = {source: markdown.split('\n')};
            const rendered = md.render(markdown, env);

            const actual = normalizeMD(rendered);

            expect(actual).toMatchSnapshot();
        });
    });
});
