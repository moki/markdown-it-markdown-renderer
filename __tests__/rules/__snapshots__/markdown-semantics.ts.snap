// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown semantics List items 253 1`] = `
"A paragraph
with two lines.

    indented code
> A block quote."
`;

exports[`markdown semantics List items 254 1`] = `
"1.  A paragraph
with two lines.

        indented code
    > A block quote."
`;

exports[`markdown semantics List items 255 1`] = `
"- one

two"
`;

exports[`markdown semantics List items 256 1`] = `
"- one
  
  two"
`;

exports[`markdown semantics List items 258 1`] = `
" -    one
      
      two"
`;

exports[`markdown semantics List items 259 1`] = `
"> >  1.  one
> >      
> >      two"
`;

exports[`markdown semantics List items 260 1`] = `
">>- one
>>
>>two"
`;

exports[`markdown semantics List items 262 1`] = `
"- foo
  
  bar"
`;

exports[`markdown semantics List items 263 1`] = `
"1.  foo
    \`\`\`
    bar
    \`\`\`
    baz
    > bam"
`;

exports[`markdown semantics List items 268 1`] = `"3. ok"`;

exports[`markdown semantics List items 272 1`] = `
"    indented code
paragraph

    more code"
`;

exports[`markdown semantics List items 273 1`] = `
"1.     indented code
   paragraph

       more code"
`;

exports[`markdown semantics List items 274 1`] = `
"1.      indented code
   paragraph

       more code"
`;

exports[`markdown semantics List items 275 1`] = `
"foo

bar"
`;

exports[`markdown semantics List items 276 1`] = `
"-    foo

bar"
`;

exports[`markdown semantics List items 277 1`] = `
"-  foo
   
   bar"
`;

exports[`markdown semantics List items 279 1`] = `
"-
  foo"
`;

exports[`markdown semantics List items 280 1`] = `
"-

foo"
`;

exports[`markdown semantics List items 282 1`] = `
"- foo
-
- bar"
`;

exports[`markdown semantics List items 286 1`] = `
" 1.  A paragraph
with two lines.

         indented code
     > A block quote."
`;

exports[`markdown semantics List items 287 1`] = `
"  1.  A paragraph
with two lines.

          indented code
      > A block quote."
`;

exports[`markdown semantics List items 288 1`] = `
"   1.  A paragraph
with two lines.

           indented code
       > A block quote."
`;

exports[`markdown semantics List items 290 1`] = `
"  1.  A paragraph
with two lines.

          indented code
      > A block quote."
`;

exports[`markdown semantics List items 291 1`] = `
"  1.  A paragraph
with two lines."
`;

exports[`markdown semantics List items 293 1`] = `
"> 1. > Blockquote
continued here."
`;

exports[`markdown semantics List items 299 1`] = `"1. - 1. foo"`;

exports[`markdown semantics Lists 306 1`] = `
"- foo
- bar
- baz"
`;

exports[`markdown semantics Lists 307 1`] = `
"- foo
  - bar
    - baz
      
      bim"
`;

exports[`markdown semantics Lists 308 1`] = `
"- foo
- bar
<!-- -->
- baz
- bim"
`;

exports[`markdown semantics Lists 309 1`] = `
"-   foo
    
    notcode
-   foo
<!-- -->

    code"
`;

exports[`markdown semantics Lists 312 1`] = `
"- a
 - b
  - c
   - d
- e"
`;

exports[`markdown semantics Lists 314 1`] = `
"- a
- b
- c"
`;

exports[`markdown semantics Lists 315 1`] = `
"* a
*
* c"
`;

exports[`markdown semantics Lists 316 1`] = `
"- a
- b
  
  c
- d"
`;

exports[`markdown semantics Lists 317 1`] = `
"- a
- b
- d"
`;

exports[`markdown semantics Lists 319 1`] = `
"- a
  - b
    
    c
- d"
`;

exports[`markdown semantics Lists 325 1`] = `
"* foo
  * bar
  
  baz"
`;

exports[`markdown semantics Lists 326 1`] = `
"- a
  - b
  - c
- d
  - e
  - f"
`;
